server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -327,13 +340,73 @@ protected void setMinSeqNoToKeep(long seqNo) {
+        for (long generation = minRemoteGenReferenced - 1; generation >= 0; generation--) {
+            if (fileTransferTracker.uploaded(Translog.getFilename(generation)) == false) {
+                break;
+            }
+            generationsToDelete.add(generation);
+        }
+        if (generationsToDelete.isEmpty() == false) {
+            deleteRemoteGeneration(generationsToDelete);
+            deleteOlderPrimaryTranslogFilesFromRemoteStore();
+        }
+    }
+
+    /**
+     * This method must be called only after there are valid generations to delete in trimUnreferencedReaders as it ensures
+     * implicitly that minimum primary term in latest translog metadata in remote store is the current primary term.
+     */
+    private void deleteOlderPrimaryTranslogFilesFromRemoteStore() {
+        // The deletion of older translog files in remote store is on best-effort basis, there is a possibility that there
+        // are older files that are no longer needed and should be cleaned up. In here, we delete all files that are part
+        // of older primary term.
+        if (olderPrimaryCleaned.trySet(Boolean.TRUE)) {
+            logger.info("Cleaning up translog uploaded by previous primaries");
+            long minPrimaryTermInMetadata = current.getPrimaryTerm();
+            Set<Long> primaryTermsInRemote = getPrimaryTermsInRemote();
+            // Delete all primary terms that are no more referenced by the metadata file and exists in the
+            Set<Long> primaryTermsToDelete = primaryTermsInRemote.stream()
+                .filter(term -> term < minPrimaryTermInMetadata)
+                .collect(Collectors.toSet());
+            primaryTermsToDelete.forEach(term -> translogTransferManager.deleteTranslogAsync(term, new ActionListener<>() {
+                @Override
+                public void onResponse(Void response) {
+                    // NO-OP
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.error(
+                        () -> new ParameterizedMessage("Exception occurred while deleting older translog files for primary_term={}", term),
+                        e
+                    );
+                }
+            }));
+        }
+    }
+
+    private Set<Long> getPrimaryTermsInRemote() {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            return translogTransferManager.listPrimaryTerms();
+        } catch (IOException e) {
+            logger.error("Exception occurred while getting primary terms from remote store", e);
         }
-        super.deleteReaderFiles(reader);
+        return LongStream.range(0, current.getPrimaryTerm()).boxed().collect(Collectors.toSet());
```

Comment:
This doesn't look the right behavior. We should propagate the exception and let the caller handle the same. 

I think it is okay to not delete older primary folders here.

Reply:
@gbbafna So, this method in RemoteFsTranslog in the caller which is asking to list all primary terms in remote. This is a private method and if someone wants to get all primary terms with in RemoteFsTranslog class, then they use `translogTransferManager.listPrimaryTerms()` which throws the exception to the caller. Currently for the deletion usecase, if there are failures in getting list of primary terms from remote then we default to clearing all primary terms which are below the minimum primary term referenced in metadata file.

Reply:
The name `getPrimaryTermsInRemote` doesn't suggest that it is going to give all previous primary terms in remote translog store. Should we just inline it then in `deleteOlderPrimaryTranslogFilesFromRemoteStore` as whatever it is returning is strictly meant for  `deleteOlderPrimaryTranslogFilesFromRemoteStore`?

Reply:
Sounds reasonable. Incorporated it.